tests/spec_decode/e2e/test_multistep_correctness.py

```diff
@@ -611,3 +611,54 @@ def test_many_k(baseline_llm_generator, test_llm_generator, batch_size: int,
                                          batch_size,
                                          max_output_len=output_len,
                                          force_output_len=True)
+
+
+@pytest.mark.parametrize(
+    "common_llm_kwargs",
+    [{
+        "model": "JackFram/llama-160m",
+
+        # Skip cuda graph recording for fast test.
+        "enforce_eager": True,
+
+        # Required for spec decode.
+        "use_v2_block_manager": True
+    }])
+@pytest.mark.parametrize("per_test_common_llm_kwargs", [{}])
+@pytest.mark.parametrize("baseline_llm_kwargs", [{}])
+@pytest.mark.parametrize(
+    "test_llm_kwargs",
+    [
+        {
+            "speculative_model": "JackFram/llama-68m",
+            "num_speculative_tokens": k,
+            "disable_bonus_tokens_in_kv_cache": False,
+        }
+        # Try small k for speculative decoding.
+        for k in [1, 2, 3]
+    ])
+@pytest.mark.parametrize("batch_size", [1, 8, 64])
+@pytest.mark.parametrize(
+    "output_len",
+    [
+        # Use smaller output len for fast test.
+        256,
+    ])
+@pytest.mark.parametrize("seed", [1])
+def test_enable_bonus_tokens_in_kv_cache(
+    baseline_llm_generator, test_llm_generator, batch_size: int,
+    output_len: int):
+    """
+    Test speculative decoding with bonus token acceptance enabled for a model 
+    using KV cache.
+
+    This test verifies that speculative decoding produces results exactly 
+    matching those from non-speculative decoding across various values of 'k' 
+    and batch sizes when bonus token acceptance is enabled. It ensures 
+    correctness by comparing the output of speculative decoding with the baseline.
+    """
+    run_greedy_equality_correctness_test(baseline_llm_generator,
```

Comment:
Test looks good. Can you manually check the draft acceptance rate? for the same draft/target model it should be 100%. without your fix (and with bonus token enabled) the acceptance rate goes to like 80%.

Ideally we have a test for this, not sure how easy that is..

Reply:
Yeah I thought of adding such an e2e test but I could find an easy way to access the metrics_collector and the stats. As you suggested I added an e2e test in test_multistep_correctness.py with both draft and target model as same ("JackFram/llama-68m") and number of speculation as 3. The System efficiency with bonus token enabled is 1 and with bonus token disabled is 0.75

**With bonus token**
INFO 07-08 06:27:27 metrics.py:316] Speculative metrics: Draft acceptance rate: 1.000, System efficiency: 1.000, Number of speculative tokens: 3, Number of accepted tokens: 21696, Number of draft tokens tokens: 21696, Number of emitted tokens tokens: 28928.

**Without bonus token**
Speculative metrics: Draft acceptance rate: 1.000, System efficiency: 0.750, Number of speculative tokens: 3, Number of accepted tokens: 21696, Number of draft tokens tokens: 21696, Number of emitted tokens tokens: 21696.